hnn_qt5.py

```diff
@@ -218,71 +231,174 @@ def runsim (self):
     print(''); self.updatewaitsimwin('')
   
   def optmodel (self):
+    import simdat
+    simdat.initial_ddat = simdat.ddat.copy()
+
     self.updatewaitsimwin('Optimizing model. . .')
-    from neuron import h # for praxis
+
+    self.last_step = False
+    self.first_step = True
+    for step in dconf['opt_info'].keys():
+      self.opt_params = dconf['opt_info'][step]
+
+      self.cur_step = step
+      if self.cur_step == len(dconf['opt_info']) - 1:
+        self.last_step = True
+        # update ranges and recalculate weights
+        self.updateoptparams()
+        sleep(1)
+
+        # reload opt_params for the last step in case the number of
+        # steps was changed by updateoptparams()
+        self.opt_params = dconf['opt_info'][len(dconf['opt_info']) - 1]
+
+      print("Starting optimization step %d"%(step+1))
+      self.runOptStep()
+
+      if not self.best_ddat is None:
+        simdat.ddat = self.best_ddat.copy()
+
+      simdat.updateoptdat(paramf,simdat.ddat['dpl']) # update optdat with best from this step
+
+      # put best opt results into GUI
+      push_values = OrderedDict()
+      for param_name in self.opt_params['ranges'].keys():
+        push_values[param_name] = self.opt_params['ranges'][param_name]['initial']
+
+      self.updatebaseparamwin(push_values)
+      sleep(1)
+
+      if self.first_step:
+        # update the plots now because we didn't do it on the first iteration
+        # for the first step
```

Comment:
TODO: I don't think this is necessary, because the display will get updated immediately after this when the second step starts.